cppcache/integration/test/PdxSerializerTest.cpp

```diff
@@ -0,0 +1,165 @@
+ public:
+  static size_t objectSize(const std::shared_ptr<const void>&,
+                           const std::string& className) {
+    EXPECT_EQ(CLASSNAME1, className);
+    auto nonPdxType = std::make_shared<NonPdxType>();
+    return sizeof(nonPdxType->longValue);
+  }
+
+  UserObjectSizer getObjectSizer(const std::string& className) override {
+    EXPECT_EQ(CLASSNAME1, className);
+    return objectSize;
+  }
+
+  std::shared_ptr<void> fromData(const std::string& className,
+                                 PdxReader& pdxReader) override {
+    EXPECT_EQ(CLASSNAME1, className);
+    auto nonPdxType = std::make_shared<NonPdxType>();
+    try {
+      nonPdxType->longValue = pdxReader.readLong("longValue");
+    } catch (...) {
+      return nullptr;
+    }
+
+    return std::move(nonPdxType);
+  }
+
+  bool toData(const std::shared_ptr<const void>& testObject,
+              const std::string& className, PdxWriter& pdxWriter) override {
+    EXPECT_EQ(CLASSNAME1, className);
+
+    auto nonPdxType = std::static_pointer_cast<const NonPdxType>(testObject);
+
+    try {
+      pdxWriter.writeLong("longValue", nonPdxType->longValue);
+      pdxWriter.markIdentityField("longValue");
+    } catch (...) {
+      return false;
+    }
+    return true;
+  }
+};
+
+std::shared_ptr<Region> setupRegion(Cache& cache) {
+  auto region = cache.createRegionFactory(RegionShortcut::PROXY)
+                    .setPoolName("default")
+                    .create("region");
+
+  return region;
+}
+
+void assertNonPdxType(const std::shared_ptr<NonPdxType>& expected,
+                      const std::shared_ptr<Cacheable>& actual) {
+  ASSERT_NE(nullptr, actual);
+  auto pdxWrapper = std::dynamic_pointer_cast<PdxWrapper>(actual);
+  ASSERT_NE(nullptr, pdxWrapper);
+  auto object = pdxWrapper->getObject();
+  ASSERT_NE(nullptr, object);
+  auto nonPdxType = std::static_pointer_cast<NonPdxType>(object);
+  ASSERT_NE(nullptr, nonPdxType);
+  EXPECT_EQ(2, nonPdxType->getLongValue());
+
+  EXPECT_NE(expected, nonPdxType);
+  EXPECT_EQ(*expected, *nonPdxType);
+}
+
+TEST(PdxSerializerTest, canSerializeNonPdxSerializableType) {
+  Cluster cluster{LocatorCount{1}, ServerCount{1}};
+  cluster.getGfsh()
+      .create()
+      .region()
+      .withName("region")
+      .withType("REPLICATE")
+      .execute();
+
+  auto nonPdxType = std::make_shared<NonPdxType>();
+  nonPdxType->setLongValue(2);
+
+  {
+    auto cache = cluster.createCache();
+    auto region = setupRegion(cache);
+    cache.getTypeRegistry().registerPdxSerializer(
+        std::make_shared<TestPdxSerializer>());
+
+    region->put("2", std::make_shared<PdxWrapper>(nonPdxType, CLASSNAME1));
+    assertNonPdxType(nonPdxType, region->get("2"));
+  }
+
+  {
+    auto cache = cluster.createCache();
+    auto region = setupRegion(cache);
+    cache.getTypeRegistry().registerPdxSerializer(
+        std::make_shared<TestPdxSerializer>());
+
+    assertNonPdxType(nonPdxType, region->get("2"));
+  }
```

Comment:
This test seems duplicate to me.  Is there something I'm missing, like the need to test with something that was already `put` in once before?

Reply:
Its doing two things. In the first block, with client A, its creating, inserting and getting the object. It is getting the object to assert that the value is properly serialized prior to the real test. The second block, client B tries to get the object without knowing anything about the type and asserts it is successful.

If there is a different structure you would like for this test please go ahead. In creating this framework I never found a way that I was completely satisfied to clearly identify two or more tracks of work on independent caches. The old framework separated them by macros and processes and it was super complicated to debug and read. I want to avoid that with this new framework.